### PR TITLE
Issue #1271 added LZWDecode compression

### LINKED ISSUE #1271: Add support for LZWDecode compression
**Please explain your intent**
Currently `fpdf2` supports 4 compression methods for images:
https://github.com/py-pdf/fpdf2/blob/2.8.0/fpdf/image_parsing.py#L35

The PDF standard specification (_cf._ [PDF32000_2008.pdf section 7.4 Filters](https://opensource.adobe.com/dc-acrobat-sdk-docs/pdfstandards/PDF32000_2008.pdf)) defines several other compression filters, including `LZWDecode`.

Recently, this filter was implemented by @MartinThoma in our companion library `pypdf`: https://github.com/py-pdf/pypdf/pull/2883

**Describe the solution you'd like**
Support `LZWDecode` for images compression.

### PR DESCRIPTION
**Pull request by opposss**
Fixes #1271   

Added support for image compression using **LZWDecode**.

**Checklist**:

- [ ] The GitHub pipeline is OK (green), 
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [x] A unit test is covering the code added / modified by this PR

- [x] This PR is ready to be merged 

- [ ] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [ ] A mention of the change is present in `CHANGELOG.md`

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/py-pdf/fpdf2/blob/master/LICENSE).

### CODE REVIEW

**Comment by opposss** [comment]
Hi @Lucas-C I added support for compression using **LZWDecode**, and I would like you to look at my code and give your opinion. I tested this filter manually by adding images compressed with **LZWDecode** to pdf files, and it seems to work correctly. After the review, if the code is ok, I could write `unit` tests.

**Comment by Lucas-C** [comment]
> Hi @Lucas-C I added support for compression using **LZWDecode**, and I would like you to look at my code and give your opinion. I tested this filter manually by adding images compressed with **LZWDecode** to pdf files, and it seems to work correctly. Local check of `pylint` and `black` showed no errors. After the review, if the code is ok, I could write `unit` tests.

Good job 👍

I'll try to review this PR very soon, today or tomorrow.

File: fpdf/image_parsing.py [hunk 1]
```diff
@@ -527,13 +532,133 @@ def __getitem__(self, tag):
     return newimgio.read(length)
 
 
+def _to_lzwdata(img, remove_slice=None, select_slice=None):
+    CLEAR_TABLE_MARKER = 256  # Special code to indicate table reset
+    EOD_MARKER = 257  # End-of-data marker
+    INITIAL_BITS_PER_CODE = 9  # Initial code bit width
+    MAX_BITS_PER_CODE = 12  # Maximum code bit width
```
Comment (Lucas-C) [inline fpdf/image_parsing.py:539]: In order to conform to the existing code style on `fpdf2`, those could be modules constants, prefixed with `LZW_`

File: fpdf/image_parsing.py [hunk 2]
```diff
@@ -527,13 +532,133 @@ def __getitem__(self, tag):
...
+
+        """
+
+        table = {bytes([i]): i for i in range(256)}
+        next_code = EOD_MARKER + 1
+        bits_per_code = INITIAL_BITS_PER_CODE
+        max_code_value = (1 << bits_per_code) - 1
+        return table, next_code, bits_per_code, max_code_value
+
+    def pack_codes_into_bytes(codes):
```
Comment (Lucas-C) [inline fpdf/image_parsing.py:573]: In order to conform to the existing code style on `fpdf2`, it's better to define those functions at the module level, an not inside the `_to_lzwdata()` function

File: fpdf/image_parsing.py [hunk 3]
```diff
@@ -527,13 +532,133 @@ def __getitem__(self, tag):
...
+        data_with_padding = data
+    else:
+        channels_count = len(data) // (img.size[0] * img.size[1])
+        row_size = img.size[0] * channels_count
+        data_with_padding = bytearray()
+        for i in range(0, len(data), row_size):
+            data_with_padding.extend(b"\0")
+            data_with_padding.extend(data[i : i + row_size])
+
+    data = data_with_padding
```
Comment (Lucas-C) [inline fpdf/image_parsing.py:559]:
```suggestion
    if img.mode == "1":
        row_size = ceil(img.size[0] / 8)
    else:
        channels_count = len(data) // (img.size[0] * img.size[1])
        row_size = img.size[0] * channels_count
        data_with_padding = bytearray()
        for i in range(0, len(data), row_size):
            data_with_padding.extend(b"\0")
            data_with_padding.extend(data[i : i + row_size])
        data = data_with_padding
```
Is padding only required when `img.mode != "1"`?

In `_to_zdata()` it seems that we always perform padding.

**Review by Lucas-C — CHANGES_REQUESTED** [review]
Hi @opposss

God job overall 👍

You placed the code at the right place, and it's clear.

I'll finish the code review once unit tests have been added, but it's a promising start!

**Comment by Lucas-C** [comment]
OK so there is one thing currently blocking in the GItHub Actions pipeline:
```
6.1.10-1 (1): test/image/image_types/image_types_insert_jpg_lzwdecode.pdf
```
This is the relevant VeraPDF rule: https://github.com/veraPDF/veraPDF-validation-profiles/wiki/PDFA-Part-1-rules/#rule-6110-1

I think the best fix is to add this rule (`6.1.10-1`) into `verapdf-ignore.json`, with reason: `fpdf2 wants to support `LZWDecode filter`

File: fpdf/image_parsing.py (hunk 3 above)
Comment (Lucas-C) [reply fpdf/image_parsing.py]: I'm not sur that you adressed that comment @opposss? 🙂

---

Could you also please a mention of this addition (the `LZWDecode` compression filter) in `CHANGELOG.md`, as part of this PR?

Comment (opposss) [reply fpdf/image_parsing.py]: Yes, you're right @Lucas-C , I changed the addition of padding, following the example with `_to_zdata()`, and also added a mention in `CHANGELOG.md` in my last commit.

**Review by Lucas-C — APPROVED** [review]

File: fpdf/image_parsing.py (hunk 3 above)
Comment (Lucas-C) [reply fpdf/image_parsing.py]: Perfect, good job 👍

**Comment by Lucas-C** [comment]
Thank you for your contribution @opposss 👍

@allcontributors please add @opposss for code

**Comment by Lucas-C** [comment]
I noticed today that the unit test `test_insert_jpg_lzwdecode` is quite slow to execute: ~78s on my computer.

And 90% of this execution is spent in `pack_codes_into_bytes()` based on this quick test:
```
pip install pytest-profiling
pytest test/image/image_types/test_insert_images.py -k lzwdecode --profile
```

I wonder if this could be improved...

**Comment by opposss** [comment]
@Lucas-C 
Apparently there is a problem with encoding large amounts of data in the case of JPEG, in case of png and other formats everything seems to work fine.

I changed the implementation `pack_codes_into_bytes()` a bit, and so far I've only managed to reduce the time from 80s to ~40s.
I will try to improve the performance even more in the coming days.

**Comment by Lucas-C** [comment]
> I changed the implementation `pack_codes_into_bytes()` a bit, and so far I've only managed to reduce the time from 80s to ~40s.
> I will try to improve the performance even more in the coming days.

Hi @opposss

Have you been able to improve this further? 🙂

**Comment by Lucas-C** [comment]
I opened issue https://github.com/py-pdf/fpdf2/issues/1380 regarding this.

**Pull request merged**